File: Estruturas_de_Dados/abstratos/Arvore/Arvore.java
package Complexo.Arvore;

import java.util.ArrayList;

public class Arvore {
    
    public static void main(String[] args) {

        BinaryTree<Integer> tree = new BinaryTree<>();

        tree.addNode(1);
        tree.addNode(2);
        tree.addNode(3);

        tree.depthSearchPreOrder(tree.root);
        tree.depthSearchInOrder(tree.root);

        tree.depthSearchPostOrder(tree.root);

        tree.breadthSearch();

        tree.addNode(4);
        tree.addNode(5);

        tree.breadthSearch();
    }

}

class Node<Thing> {

    Thing data;
    Node<Thing> left;
    Node<Thing> right;

    Node(Thing data) {
        this.data = data;
    }

}

class BinaryTree<Thing> {

    Node<Thing> root;

    void printNodeTree(Node<Thing> node) {

        System.out.print("| Node Value: " + node.data);

        if (node.left == null) {
            System.out.print(" | Left: null");
        } else {
            System.out.print(" | Left: " + node.left.data);
        }

        if (node.right == null) {
            System.out.print(" | Right: null |");
        } else {
            System.out.print(" | Right: " + node.right.data);
        }

        System.out.println(" |");
    }

    void addNode(Thing data) {

        if (this.root == null) {
            this.root = new Node<>(data);
            return;
        }

        ArrayList<Node<Thing>> queue = new ArrayList<>();

        queue.add(this.root);
        
        while (!queue.isEmpty()) {

            Node<Thing> node = queue.getFirst();
            queue.removeFirst();

            if (node.left == null) {
                node.left = new Node<>(data);
                return;
            } else if (node.right == null) {
                node.right = new Node<>(data);
                return;
            } else {
                queue.add(node.left);
                queue.add(node.right);
            }

        }
        
    }

    void depthSearchPreOrder (Node<Thing> node) {

        if (this.root == null) {
            System.err.println("\nA arvore esta vazia");
        }

        if (node == null) {
            node = this.root;
        }

        System.out.println("\nDepth Search Pre Order");

        this.printNodeTree(node);

        if (node.left != null) {
            this.depthSearchPreOrder(node.left);
        }
        if (node.right != null) {
            this.depthSearchPreOrder(node.right);
        }
    }

    void depthSearchInOrder (Node<Thing> node) {

        if (this.root == null) {
            System.err.println("\nA arvore esta vazia");
        }

        if (node == null) {
            node = this.root;
        }

        System.out.println("\nDepth Search In Order");

        if (node.left != null) {
            this.depthSearchInOrder(node.left);
        }

        this.printNodeTree(node);

        if (node.right != null) {
            this.depthSearchInOrder(node.right);
        }
    }

    void depthSearchPostOrder (Node<Thing> node) {

        if (this.root == null) {
            System.err.println("\nA arvore esta vazia");
        }

        if (node == null) {
            node = this.root;
        }

        System.out.println("\nDepth Search Post Order");

        if (node.left != null) {
            this.depthSearchPostOrder(node.left);
        }
        if (node.right != null) {
            this.depthSearchPostOrder(node.right);
        }

        this.printNodeTree(node);

    }

    void breadthSearch() {

        if (this.root == null) {
            System.err.println("\nA arvore esta vazia");
            return;
        }

        
        ArrayList<Node<Thing>> queue = new ArrayList<>();
        
        queue.add(this.root);

        System.out.println("\nBreadth Search");

        while (!queue.isEmpty()) {

            Node<Thing> node = queue.getFirst();
            queue.removeFirst();

            if (node.left != null) {
                queue.add(node.left);
            } 
            if (node.right != null) {
                queue.add(node.right);
            }

            this.printNodeTree(node);
            
            
        }
    }
}

File: Estruturas_de_Dados/abstratos/Arvore/arvore.py
# É meio óbvio falar isso, mas na árvore,
# os nodes possuem 3 variavies: o dado, a esquerda e a direita.
# Essas duas últimas, são como a variavel próximo
# das estruturas anteriores. E apesar do nome diferente, age
# da mesma forma: armazenando qualquer node que você botar ali

class Node():
    def __init__(self, value):
        
        self.value = value
        self.left = None
        self.right = None
    
    def __repr__(self):
        return f"{self.value}"

class Binary_Tree():
    def __init__(self, root:Node = None):
        
        self.root = root 
    
    def print_node_tree(self, node:Node):
        print(f"| Node Value: {node.value} | Left: {node.left} | Right: {node.right} |")

    def add_node(self, value):
        
        # Essa função básicamente usa uma função BFS (mais detalhes abaixo)
        # pra procurar um espaço vázio e colocar um node lá, priorizando 
        # sempre a esquerda

        # Eu esqueci de escreve essa função, mas um código para removes nodes
        # funcionária da mesma forma

        if not self.root:
            self.root = Node(value)
            print(f"Root of tree was created with value {self.root.value}")
            return
        
        queue = []
        queue.append(self.root)

        while queue:

            node = queue.pop(0)

            if not node.left:
                node.left = Node(value)
                return
            elif not node.right:
                node.right = Node(value)
                return
            else:
                queue.append(node.left)
                queue.append(node.right)

    def depth_search_preorder(self, node = None):

        if node is None:
            node = self.root
        
        if not node:
            print("A arvre ta vazia")
            return
    
        #print(node.value)
        self.print_node_tree(node)
                
        if node.left:
            self.depth_search_preorder(node.left)    
        if node.right:
            self.depth_search_preorder(node.right)
               
    def depth_search_inorder(self, node = None):
        
        if node is None:
            node = self.root
            
        if not node:
            print("A arvre ta vazia")
            return
        
        
        if node.left:
            self.depth_search_inorder(node.left)    
            
        #print(node.value)
        self.print_node_tree(node)
        
        if node.right:
            self.depth_search_inorder(node.right)

    def depth_search_postorder(self, node = None):
        
        if node is None:
            node = self.root
        if not node:
            print("A arvre ta vazia")
            return
        
        
        if node.left:
            self.depth_search_postorder(node.left)    
        if node.right:
            self.depth_search_postorder(node.right)
            
        #print(node.value)
        self.print_node_tree(node)

    def breadth_search(self):
        
        # Breadth Search é uma função de busca assim como DFS,
        # mas diferente das funções acima, aqui o código armazena
        # cada node num array (ou fila se preferir)

        # O código roda num While Loop, e continua nesse loop
        # até que essa fila esteja vazio.

        # Por conta dos primeiros visitados serem os mais próxmos
        # do node atual, isso faz com que o BFS faça 
        # uma varredura em níveis da árvore, ao invés de ir fundo
        # em cada galho, como no DFS

        queue = []

        if not self.root:
            return
        
        node = self.root
        queue.append(node)

        while queue:

            node = queue.pop(0)

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
            
            #print(node.value)
            self.print_node_tree(node)
        
        print(end="\n\n")

#        f
#       / \
#      b   g
#     / \    \
#   a   d    i
#       / \   /
#      c   e h

if __name__ == "__main__":

    bin_tree = Binary_Tree()

    bin_tree.add_node(1)
    bin_tree.add_node(3)

    print("Depth Pre Search:")
    bin_tree.depth_search_preorder()

    bin_tree.add_node(5)
    bin_tree.add_node(7)
    bin_tree.add_node(9)

    print("\nDepth In Search:")
    bin_tree.depth_search_inorder()

    bin_tree.add_node(11)
    bin_tree.add_node(13)
    bin_tree.add_node(15)

    print("\nDepth Post Search:")
    bin_tree.depth_search_postorder()

    bin_tree.add_node(17)
    bin_tree.add_node(19)
    bin_tree.add_node(21)

    print("\nBreadth Search:")
    bin_tree.breadth_search()